### PR TITLE
Easily enable apex O1 amp from train_dalle.py

### PR DESCRIPTION
**Pull request by afiaka87**
CogView sidestepped needing to "tame" 16 bit precision by just using "O2" (almost 16 bit) precision. 

I tried implementing O2 and I think it's rather possible but there are some casts from Long to Float which need to happen for it to work. 

The O1 case works very well as is however. In my case (RTX 2070) - I go from ~16 samples per second to ~40 samples per second. using fp16 instead I get around ~50 samples. So it's faster but not necessarily by much. I assume this benchmark changes drastically when the comms latency of a multi-node system needs to be considered and my intuition is that sending half as many bits across the wire can make something like optimizer offloading and allreduce more effective. 

tl;dr automatic mixed precision is slower than fp16, but you'll be able to see generations during training and won't have to deal with potential divergence issues caused by training fp16.

### CODE REVIEW

**Comment by afiaka87** [comment]
In order to run with Apex AMP (through DeepSpeed), you will need to install DeepSpeed using either the Dockerfile or the bash script. Then you will need to install apex from source:

```sh
git clone https://github.com/NVIDIA/apex
cd apex
pip install -v --disable-pip-version-check --no-cache-dir --global-option="--cpp_ext" --global-option="--cuda_ext" ./
cd ..
```

In order to train with apex automatic mixed precision, you can do something like this:
```sh
deepspeed train_dalle.py --taming --image_text_folder Datasets --depth 8 --heads 8 --distr_backend deepspeed --amp
```

Or the shorter method:
```sh
deepspeed train_dalle.py --taming --image_text_folder Datasets --depth 8 --heads 8 --deepspeed --amp
```

**Review by rom1504 — APPROVED** [review]
that's great
